client/allocrunner/taskrunner/secrets_hook.go

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: BUSL-1.1
+
+package taskrunner
+
+import (
+	"context"
+	"fmt"
+	"maps"
+	"path/filepath"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/client/allocrunner/interfaces"
+	ti "github.com/hashicorp/nomad/client/allocrunner/taskrunner/interfaces"
+	"github.com/hashicorp/nomad/client/allocrunner/taskrunner/secrets"
+	"github.com/hashicorp/nomad/client/allocrunner/taskrunner/template"
+	"github.com/hashicorp/nomad/client/config"
+	"github.com/hashicorp/nomad/client/taskenv"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// Currently only support Vault and Nomad which use CT. Future work
+// can modify this interface to include custom providers using
+// a plugin interface.
+type SecretProvider interface {
+	// BuildTemplates should construct a template appropriate for
+	// that provider and append it to the templateManager's templates.
+	BuildTemplate() (*structs.Template, error)
+
+	// Each provider implementation much be able to parse it's "response" object.
+	Parse() (map[string]string, error)
+}
+
+type secretsHookConfig struct {
+	// logger is used to log
+	logger log.Logger
+
+	// lifecycle is used to interact with the task's lifecycle
+	lifecycle ti.TaskLifecycle
+
+	// events is used to emit events
+	events ti.EventEmitter
+
+	// clientConfig is the Nomad Client configuration
+	clientConfig *config.Config
+
+	// envBuilder is the environment variable builder for the task.
+	envBuilder *taskenv.Builder
+
+	// nomadNamespace is the job's Nomad namespace
+	nomadNamespace string
+}
+
+type secretsHook struct {
+	// logger is used to log
+	logger log.Logger
+
+	// lifecycle is used to interact with the task's lifecycle
+	lifecycle ti.TaskLifecycle
+
+	// events is used to emit events
+	events ti.EventEmitter
+
+	// clientConfig is the Nomad Client configuration
+	clientConfig *config.Config
+
+	// envBuilder is the environment variable builder for the task
+	envBuilder *taskenv.Builder
+
+	// nomadNamespace is the job's Nomad namespace
+	nomadNamespace string
+
+	// secrets to be fetched and populated for interpolation
+	secrets []*structs.Secret
+
+	// taskrunner secrets map
+	taskSecrets map[string]string
+}
+
+func newSecretsHook(conf *secretsHookConfig, secrets []*structs.Secret) *secretsHook {
+	return &secretsHook{
+		logger:         conf.logger,
+		lifecycle:      conf.lifecycle,
+		events:         conf.events,
+		clientConfig:   conf.clientConfig,
+		envBuilder:     conf.envBuilder,
+		nomadNamespace: conf.nomadNamespace,
+		secrets:        secrets,
+		taskSecrets:    make(map[string]string),
+	}
+}
+
+func (h *secretsHook) Name() string {
+	return "secrets"
+}
+
+func (h *secretsHook) Prestart(ctx context.Context, req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
+	providers, templates := []SecretProvider{}, []*structs.Template{}
+	for idx, s := range h.secrets {
+		tmplPath := filepath.Join(req.TaskDir.SecretsDir, fmt.Sprintf("%d", idx))
```

Comment:
It might be a good idea to name explicitly as a tempfile.

```suggestion
		tmplPath := filepath.Join(req.TaskDir.SecretsDir, fmt.Sprintf("temp-%d", idx))
```

But I like that we're not allowing the user to control the filepath here at all, as that's been a source of issues in the past. And because this is a prestart task we don't have much exposure to symlink attacks (except on task restart, which the template manager already has protections for). :+1: